File: db4on/reference/vb/db4o-reference-chapters/src/com/db4odoc/f1/staticfields/StaticFieldExample.vb
' Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com
Imports System
Imports System.IO
Imports com.db4o
Imports com.db4o.query

Namespace com.db4odoc.f1.staticfields
    Public Class StaticFieldExample
        Inherits Util

        Public Sub New()
        End Sub

        Public Shared Sub main(ByVal args() As String)
            SetPilotsSimple()
            CheckPilots()

            SetPilotsStatic()
            CheckPilots()
            UpdatePilots()
            UpdatePilotCategories()
            CheckPilots()
            DeleteTest()

        End Sub


        Public Shared Sub SetPilotsSimple()
            Console.WriteLine("In the default setting, static constants are not continously stored and updated.")
            File.Delete(Util.YapFileName)
            Dim db As ObjectContainer = Db4o.OpenFile(Util.YapFileName)
            Try
                db.Set(New Pilot("Michael Schumacher", PilotCategories.WINNER))
                db.Set(New Pilot("Rubens Barrichello", PilotCategories.TALENTED))
            Finally
                db.Close()
            End Try
        End Sub

        Public Shared Sub SetPilotsStatic()
            Console.WriteLine("The feature can be turned on for individual classes.")
            Db4o.Configure().ObjectClass(GetType(PilotCategories)).PersistStaticFieldValues()
            File.Delete(Util.YapFileName)
            Dim db As ObjectContainer = Db4o.OpenFile(Util.YapFileName)
            Try
                db.Set(New Pilot("Michael Schumacher", PilotCategories.WINNER))
                db.Set(New Pilot("Rubens Barrichello", PilotCategories.TALENTED))
            Finally
                db.Close()
            End Try
        End Sub

        Public Shared Sub CheckPilots()
            Dim db As ObjectContainer = Db4o.OpenFile(Util.YapFileName)
            Try
                Dim result As ObjectSet = db.Get(GetType(Pilot))
                Dim x As Integer
                For x = 0 To result.Count - 1 Step x + 1
                    Dim pilot As Pilot = CType(result(x), Pilot)
                    If pilot.Category Is PilotCategories.WINNER Then
                        Console.WriteLine("Winner pilot: " + pilot.ToString())
                    ElseIf pilot.Category Is PilotCategories.TALENTED Then
                        Console.WriteLine("Talented pilot: " + pilot.ToString())
                    Else
                        Console.WriteLine("Uncategorized pilot: " + pilot.ToString())
                    End If
                Next
            Finally
                db.Close()
            End Try
        End Sub

        Public Shared Sub UpdatePilots()
            Console.WriteLine("Updating PilotCategory in pilot reference:")
            Dim db As ObjectContainer = Db4o.OpenFile(Util.YapFileName)
            Try
                Dim result As ObjectSet = db.Get(GetType(Pilot))
                Dim x As Integer
                For x = 0 To result.Count - 1 Step x + 1
                    Dim pilot As Pilot = CType(result(x), Pilot)
                    If pilot.Category Is PilotCategories.WINNER Then
                        Console.WriteLine("Winner pilot: " + pilot.ToString())
                        Dim pc As PilotCategories = pilot.Category
                        pc.TestChange("WINNER2006")
                        db.Set(pilot)
                    End If
                Next
            Finally
                db.Close()
            End Try
            PrintCategories()
        End Sub

        Public Shared Sub UpdatePilotCategories()
            Console.WriteLine("Updating PilotCategories explicitly:")
            Dim db As ObjectContainer = Db4o.OpenFile(Util.YapFileName)
            Try
                Dim result As ObjectSet = db.Get(GetType(PilotCategories))
                Dim x As Integer
                For x = 0 To result.Count - 1 Step x + 1
                    Dim pc As PilotCategories = CType(result(x), PilotCategories)
                    If pc Is PilotCategories.WINNER Then
                        pc.TestChange("WINNER2006")
                        db.Set(pc)
                    End If
                Next
            Finally
                db.Close()
            End Try
            PrintCategories()
        End Sub

        Public Shared Sub DeleteTest()
            Dim db As ObjectContainer = Db4o.OpenFile(Util.YapFileName)
            db.Ext().Configure().ObjectClass(GetType(Pilot)).CascadeOnDelete(True)
            Try
                Console.WriteLine("Deleting Pilots :")
                Dim result As ObjectSet = db.Get(GetType(Pilot))
                Dim x As Integer
                For x = 0 To result.Count - 1 Step x + 1
                    Dim pilot As Pilot = CType(result(x), Pilot)
                    db.Delete(pilot)
                Next
                PrintCategories()
                Console.WriteLine("Deleting PilotCategories :")
                result = db.Get(GetType(PilotCategories))
                For x = 0 To result.Count - 1 Step x + 1
                    db.Delete(result(x))
                Next
                PrintCategories()
            Finally
                db.Close()
            End Try
        End Sub

        Public Shared Sub PrintCategories()
            Dim db As ObjectContainer = Db4o.OpenFile(Util.YapFileName)
            Try
                Dim result As ObjectSet = db.Get(GetType(PilotCategories))
                Console.WriteLine("Stored categories: " + (result.Count).ToString())
                Dim x As Integer
                For x = 0 To result.Count - 1 Step x + 1
                    Dim pc As PilotCategories = CType(result(x), PilotCategories)
                    Console.WriteLine("Category: " + pc.ToString())
                Next
            Finally
                db.Close()
            End Try
        End Sub

        Public Shared Sub DeletePilotCategories()
            PrintCategories()
            Dim db As ObjectContainer = Db4o.OpenFile(Util.YapFileName)
            Try
                Dim result As ObjectSet = db.Get(GetType(PilotCategories))
                Dim x As Integer
                For x = 0 To result.Count - 1 Step x + 1
                    Dim pc As PilotCategories = CType(result(x), PilotCategories)
                    db.Delete(pc)
                Next
            Finally
                db.Close()
            End Try
            PrintCategories()
        End Sub
    End Class
End Namespace

File: db4on/reference/db4o-reference-chapters/src/com/db4odoc/f1/staticfields/StaticFieldExample.cs
using System;
using System.IO;
using com.db4odoc.f1;
using com.db4o;
using com.db4o.query;
using System.Drawing;

namespace com.db4odoc.f1.staticfields
{
	public class StaticFieldExample
	{
		public readonly static string YapFileName = "formula1.yap";

		public StaticFieldExample()
		{
		}
		public static void Main(String[] args) 
		{
			SetPilotsSimple();
			CheckPilots();
			//
			SetPilotsStatic();
			CheckPilots();
			UpdatePilots();
			UpdatePilotCategories();
			CheckPilots();
			DeleteTest();
		}
		// end Main

		public static void SetCar()
		{
			ObjectContainer db=Db4o.OpenFile(YapFileName);
			try 
			{
				Car car = new Car();
				car._color = Color.Green;
				db.Set(car);
			} 
			finally 
			{
				db.Close();
			}
		}
		// end SetCar

		public static void SetPilotsSimple()
		{
			Console.WriteLine("In the default setting, static constants are not continously stored and updated.");
			File.Delete(YapFileName);
			ObjectContainer db=Db4o.OpenFile(YapFileName);
			try 
			{
				db.Set(new Pilot("Michael Schumacher",PilotCategories.WINNER));
				db.Set(new Pilot("Rubens Barrichello",PilotCategories.TALENTED));
			} 
			finally 
			{
				db.Close();
			}
		}
		// end SetPilotsSimple
	
		public static void SetPilotsStatic()
		{
			Console.WriteLine("The feature can be turned on for individual classes.");
			Db4o.Configure().ObjectClass(typeof(PilotCategories)).PersistStaticFieldValues();
			File.Delete(YapFileName);
			ObjectContainer db=Db4o.OpenFile(YapFileName);
			try 
			{
				db.Set(new Pilot("Michael Schumacher",PilotCategories.WINNER));
				db.Set(new Pilot("Rubens Barrichello",PilotCategories.TALENTED));
			} 
			finally 
			{
				db.Close();
			}
		}
		// end SetPilotsStatic
	
		public static void CheckPilots()
		{
			ObjectContainer db=Db4o.OpenFile(YapFileName);
			try 
			{
				ObjectSet result = db.Get(typeof(Pilot));
				for(int x = 0; x < result.Count; x++)
				{
					Pilot pilot = (Pilot )result[x];
					if (pilot.Category == PilotCategories.WINNER)
					{
						Console.WriteLine("Winner pilot: " + pilot);
					} 
					else if (pilot.Category == PilotCategories.TALENTED)
					{
						Console.WriteLine("Talented pilot: " + pilot);
					}  
					else 
					{
						Console.WriteLine("Uncategorized pilot: " + pilot);
					}
				}
			} 
			finally 
			{
				db.Close();
			}
		}
		// end CheckPilots
	
		public static void UpdatePilots()
		{
			Console.WriteLine("Updating PilotCategory in pilot reference:");
			ObjectContainer db=Db4o.OpenFile(YapFileName);
			try 
			{
				ObjectSet result = db.Get(typeof(Pilot));
				for(int x = 0; x < result.Count; x++)
				{
					Pilot pilot = (Pilot )result[x];
					if (pilot.Category == PilotCategories.WINNER)
					{
						Console.WriteLine("Winner pilot: " + pilot);
						PilotCategories pc = pilot.Category;
						pc.TestChange("WINNER2006");
						db.Set(pilot);
					}
				}
			} 
			finally 
			{
				db.Close();
			}
			PrintCategories();
		}
		// end UpdatePilots
	
		public static void UpdatePilotCategories()
		{
			Console.WriteLine("Updating PilotCategories explicitly:");
			ObjectContainer db=Db4o.OpenFile(YapFileName);
			try 
			{
				ObjectSet result = db.Get(typeof(PilotCategories));
				for(int x = 0; x < result.Count; x++)
				{
					PilotCategories pc = (PilotCategories)result[x];
					if (pc == PilotCategories.WINNER)
					{
						pc.TestChange("WINNER2006");
						db.Set(pc);
					}
				}
			} 
			finally 
			{
				db.Close();
			}
			PrintCategories();
		}
		// end UpdatePilotCategories
	
		public static void DeleteTest()
		{
			ObjectContainer db=Db4o.OpenFile(YapFileName);
			db.Ext().Configure().ObjectClass(typeof(Pilot)).CascadeOnDelete(true);
			try 
			{
				Console.WriteLine("Deleting Pilots :");
				ObjectSet result = db.Get(typeof(Pilot));
				for(int x = 0; x < result.Count; x++)
				{
					Pilot pilot = (Pilot )result[x];
					db.Delete(pilot);
				}
				PrintCategories();
				Console.WriteLine("Deleting PilotCategories :");
				result = db.Get(typeof(PilotCategories));
				for(int x = 0; x < result.Count; x++)
				{
					db.Delete(result[x]);
				}
				PrintCategories();
			} 
			finally 
			{
				db.Close();
			}
		}
		// end DeleteTest
	
		public static void PrintCategories()
		{
			ObjectContainer db=Db4o.OpenFile(YapFileName);
			try 
			{
				ObjectSet  result = db.Get(typeof(PilotCategories));
				Console.WriteLine("Stored categories: " + result.Count);
				for(int x = 0; x < result.Count; x++)
				{
					PilotCategories pc = (PilotCategories)result[x];
					Console.WriteLine("Category: "+pc);
				}
			} 
			finally 
			{
				db.Close();
			}
		}
		// end PrintCategories
	
		public static void DeletePilotCategories()
		{
			PrintCategories();
			ObjectContainer db=Db4o.OpenFile(YapFileName);
			try 
			{
				ObjectSet  result = db.Get(typeof(PilotCategories));
				for(int x = 0; x < result.Count; x++)
				{
					PilotCategories pc = (PilotCategories)result[x];
					db.Delete(pc);
				}
			} 
			finally 
			{
				db.Close();
			}
			PrintCategories();
		}
		// end DeletePilotCategories
	}
}
